src/crimson/os/seastore/extent_placement_manager.cc

```diff
@@ -0,0 +1,269 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:nil -*-
+// vim: ts=8 sw=2 smarttab expandtab
+
+#include "crimson/os/seastore/journal.h"
+#include "crimson/os/seastore/extent_placement_manager.h"
+#include "crimson/os/seastore/lba_manager/btree/lba_btree_node_impl.h"
+#include "crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.h"
+#include "crimson/os/seastore/onode_manager/staged-fltree/node_extent_manager/seastore.h"
+#include "crimson/os/seastore/collection_manager/collection_flat_node.h"
+#include "crimson/os/seastore/object_data_handler.h"
+
+namespace {
+  seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_seastore);
+  }
+}
+
+namespace crimson::os::seastore {
+
+SegmentedAllocator::SegmentedAllocator(
+  SegmentProvider& sp,
+  SegmentManager& sm,
+  LBAManager& lba_manager,
+  Journal& journal)
+  : segment_provider(sp),
+    segment_manager(sm),
+    lba_manager(lba_manager),
+    journal(journal)
+{
+  std::generate_n(
+    std::back_inserter(writers),
+    crimson::common::get_conf<uint64_t>(
+      "seastore_init_rewrite_segments_num_per_device"),
+    [&] {
+      return Writer{
+	segment_provider,
+	segment_manager,
+	lba_manager,
+	journal};
+      });
+}
+
+SegmentedAllocator::Writer::finish_record_ret
+SegmentedAllocator::Writer::finish_write(
+  Transaction& t,
+  ool_record_t& record) {
+  return trans_intr::do_for_each(record.get_extents(),
+    [this, &t](auto& ool_extent) {
+    auto& lextent = ool_extent->get_lextent();
+    logger().debug("SegmentedAllocator::Writer::finish_write: extent: {}", *lextent);
+    return lba_manager.update_mapping(
+      t,
+      lextent->get_laddr(),
+      lextent->get_paddr(),
+      ool_extent->get_ool_paddr()
+    ).si_then([&ool_extent, &t, &lextent] {
+      ool_extent->persist_paddr();
+      lextent->backend_type = device_type_t::NONE;
+      lextent->hint = {};
+      t.add_fresh_extent(lextent, delay_inline_ool_t::DELAYED_OOL);
```

Comment:
This violates the Cache interface boundary.  Please create a cache method for this operation.  Also, don't call add_fresh_extent to *update* the address for the extent -- create another method.

In fact, I'm not sure DELAYED_OOL or DELAYED_INLINE should be elements of delay_inline_ool_t.  Is it ever valid to call alloc_new_extent with DELAYED_OOL or DELAYED_INLINE?  Those two values are only ever used here and in delayed_alloc_or_ool_write.  I think you should remove DELAYED_OOL and DELAYED_INLINE from alloc_new_extent and add_fresh_extent and add two new methods Cache::mark_delayed_extent_ool and Cache::mark_delayed_extent_inline.

Reply:
ok